src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +249,522 @@ otError DuaManager::Store(void)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    mIsDuaPending = false;
+
+    if (aResult == OT_ERROR_RESPONSE_TIMEOUT)
+    {
+        UpdateCheckDelay(Mle::KResponseTimeoutDelay);
+        ExitNow();
+    }
+
+    VerifyOrExit(aResult == OT_ERROR_NONE && aMessage.GetCode() == OT_COAP_CODE_CHANGED, OT_NOOP);
+
+    ProcessDuaResponse(aMessage);
+
+exit:
+    mRegistrationTask.Post();
+}
+
+void DuaManager::HandleDuaNotification(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    otLogInfoDua("Received DUA.ntf");
+
+    VerifyOrExit(aMessage.GetCode() == OT_COAP_CODE_POST, OT_NOOP);
+
+    if (aMessage.IsConfirmable())
+    {
+        if (Get<Coap::Coap>().SendEmptyAck(aMessage, aMessageInfo) == OT_ERROR_NONE)
+        {
+            otLogInfoDua("Sent DUA.ntf acknowledgment");
+        }
+    }
+
+    ProcessDuaResponse(aMessage);
+
+exit:
+    return;
+}
+
+void DuaManager::ProcessDuaResponse(Coap::Message &aMessage)
+{
+    Ip6::Address target;
+    uint8_t      status;
+
+    SuccessOrExit(Tlv::FindUint8Tlv(aMessage, ThreadTlv::kStatus, status));
+    SuccessOrExit(Tlv::FindTlv(aMessage, ThreadTlv::kTarget, &target, sizeof(target)));
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (Get<ThreadNetif>().HasUnicastAddress(target))
+    {
+        switch (static_cast<ThreadStatusTlv::DuaStatus>(status))
+        {
+        case ThreadStatusTlv::kDuaSuccess:
+            mLastRegistrationTime = TimerMilli::GetNow();
+            break;
+        case ThreadStatusTlv::kDuaReRegister:
+            mDuaState                  = kToRegister;
+            mDelay.mFields.mCheckDelay = Mle::kImmediateReRegisterDelay;
+            break;
+        case ThreadStatusTlv::kDuaInvalid:
+            // Domain Prefix might be invalid.
+            RemoveDomainUnicastAddress();
+            break;
+        case ThreadStatusTlv::kDuaDuplicate:
+            RemoveDomainUnicastAddress();
+            mDadCounter++;
+
+            if (GenerateDomainUnicastAddressIid() == OT_ERROR_NONE)
+            {
+                AddDomainUnicastAddress();
+            }
+
+            break;
+        case ThreadStatusTlv::kDuaNoResources:
+            // fall through
+        case ThreadStatusTlv::kDuaNotPrimary:
+            // fall through
+        case ThreadStatusTlv::kDuaGeneralFailure:
+            UpdateReregistrationDelay();
+            break;
+        }
+    }
+    else
+#endif
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    {
+        Child *child = Get<ChildTable>().GetChildAtIndex(mChildIndexDuaRegistering);
+
+        VerifyOrExit(child != NULL, OT_NOOP);
+        VerifyOrExit(child->HasIp6Address(target), OT_NOOP);
+
+        switch (status)
+        {
+        case ThreadStatusTlv::kDuaSuccess:
+            // Mark as Registered
+            mChildDuaRegisterMask.Set(mChildIndexDuaRegistering);
+            break;
+        case ThreadStatusTlv::kDuaReRegister:
+            mRegisterCurrentChildIndex = true;
```

Comment:
I am not clear about the intention of `mRegisterCurrentChildIndex`. 
This seems never changed to false (except when prefix changed). 
Is it used to indicate if `mChildIndexDuaRegistering` is actually registering rather than being initialized to 0?

How do we move to the next Child if `mRegisterCurrentChildIndex` is not set to `false`?

Reply:
good catch, will update to reset it before the `switch()`